sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -174,8 +106,110 @@ public void executeDdl(String sqlStatement) throws ParseException {
   public String explain(String sqlString) throws ParseException {
     try {
       return RelOptUtil.toString(planner.convertToBeamRel(sqlString));
-    } catch (ValidationException | RelConversionException | SqlParseException e) {
+    } catch (Exception e) {
       throw new ParseException("Unable to parse statement", e);
     }
   }
+
+  /** BeamSqlEnv's Builder. */
+  public static class BeamSqlEnvBuilder {
+    private final JdbcConnection jdbcConnection;
+    private String plannerName = "CalcitePlanner";
+
+    public static BeamSqlEnvBuilder builder(TableProvider tableProvider) {
+      return new BeamSqlEnvBuilder(tableProvider);
+    }
+
+    private BeamSqlEnvBuilder(TableProvider tableProvider) {
+      jdbcConnection = JdbcDriver.connect(tableProvider);
```

Comment:
I think that connection logic (and any other complicated logic) should live outside of the builder constructor or setter methods (e.g. applies to registering the udfs), either in `BeamSqlEnv` or at least in `build()`, this way you can pass around a not completely initialized builder to configure it, and then connect only when you know you have all the data. I.e. should be a simple data structure that only does complicated things in `build()` to pass them to the constructor of the outer class.